Updated downsampling

kbrose opened the pull request:
## PR Summary

Updates the downsampling of large lines. Specifically, this PR aims to

1. Update `Line2D` and `Path` to actually downsample when the RC parameter indicates it should. (Addressing #8775.)
2. Update the downsampling algorithm to handle (almost) anti-parallel lines as well as (almost) parallel lines. (Similar effort to #7632, but updates the C++ algorithm instead of rolling a new one in Python (@anntzer))

<!--If it fixes an open issue, please link to the issue here.-->

## PR Checklist

- [x] Has Pytest style unit tests
- [x] Code is PEP 8 compliant 
- [x] New features are documented, with examples if plot related
- [x] Documentation is sphinx and numpydoc compliant
- [x] Added an entry to doc/users/whats_new.rst if major new feature
- [x] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

anntzer commented on lib/matplotlib/lines.py:
```diff
@@ -790,6 +790,9 @@ def draw(self, renderer):
                 else:
+                    print(tpath.vertices.size)
+                    tpath = affine.inverted().transform_path(affine.transform_path(tpath).cleaned(simplify=True))
+                    print(tpath.vertices.size)
```
I guess what is happening is that you need to do the simplification in screen units, but the later draw routine needs the path in data units, so you need to convert one extra time back and forth?  That seems inefficient, any chance you could either 1. do the simplification at the last minute (once you don't need to convert back and forth anymore), or at least 2. keep the simplified path in screen units, and set the transform to identity?

Also, instead of prints (which need to go anyways :-)), some tests (e.g. with paths for which you know what the result of simplification should be) would be welcome, to show that the simplifier is indeed working as expected.

kbrose replied:
Copying a comment that I haven't pushed yet:

```python
# the path simplification code expects the path
# to be in pixel space, not data space, but the
# renderer downstream expects the path to be
# in data space. The transformation `affine`
# will put the path into pixel space, and
# affine.inverted() will transform it back to
# data space. So, transform using affine, simplify
# the path, and transform using the affine inverted.
```

It is inefficient, but I see no way around it at the moment. Do you know what populates the `renderer` argument? The only way to delay cleaning until it has been transformed is to do it inside this call: `renderer.draw_path(gc, path, trans)` from

```python
    def _draw_solid(self, renderer, gc, path, trans):
        gc.set_linestyle('solid')
        gc.set_dashes(self._dashOffset, self._dashSeq)
        renderer.draw_path(gc, path, trans)
```

kbrose replied:
Obviously print statements will be removed and tests will be added.

Testing in general will be a huge pain, since nearly every single image will fail the current image comparisons. We may have to change the default value of `path.simplify` to False to avoid this...

kbrose replied:
Yeah, it looks like `renderer` is basically the backend. Without requiring changes to the backend API (so that it takes in an already transformed path, and I'm sure it was not designed that way initially also for performance reasons), I don't think we can avoid it. It's still faster than *not* doing the downsampling.

anntzer replied:
You may be able to get away with editing _classic_test.mplstyle, which is basically there for that reason (we try not to update all test images at once if possible, as this really leads to bloating of the repo size).

kbrose replied:
@anntzer are you saying if I edit `_classic_test.mplstyle` to turn path simplification off, that all the figures created during testing will be made with simplification turned off?

kbrose replied:
Do you know if that behavior is documented anywhere? I've been reading through developer documentation but not finding anything.

anntzer replied:
I hope so, at least...

anntzer commented on lib/matplotlib/lines.py:
```diff
@@ -790,8 +790,21 @@ def draw(self, renderer):
+                        # the path, and transform using the affine inverted.
+                        tpath = affine.inverted().transform_path(
+                            affine.transform_path(tpath).cleaned(simplify=True)
+                        )
```
would something like
```
tpath = affine.transform_path(tpath).cleaned(simplify=True)
affine = IdentityTransform() # or whatever it is called
```
not work to avoid the back and forth discussed earlier?

kbrose replied:
I'll play around with that. I'm a little worried about changing that since I don't know what downstream rendering code relies on (does it use the data-space to generate axes labels? scale text? etc.). It might be premature optimization at this point.

anntzer replied:
I really meant something like "Check if affine is used anywhere after.  If not, do /previous message/.  If it is, set a flag somewhere, and only use an identity affine when drawing the path".
Well perhaps the second one *would be* an premature optimization if needed, but please give the first one a try if possible, given that that would be fairly simple.

kbrose replied:
Hmm. Initial tests just created a blank figure when doing that.

kbrose replied:
Sorry didn't see your 2nd message before posting. So I guess it looks like `affine` is being used for something else later on, I'm guessing in the clipping routine:

```<path_converters.h>
/*
 This file contains a number of vertex converters that modify
 paths. They all work as iterators, where the output is generated
 on-the-fly, and don't require a copy of the full data.

 Each class represents a discrete step in a "path-cleansing" pipeline.
 They are currently applied in the following order in the Agg backend:

   ...

   3. PathClipper: Clips line segments to a given rectangle.  This is
      helpful for data reduction, and also to avoid a limitation in
      Agg where coordinates can not be larger than 24-bit signed
      integers.

   ...

   5. PathSimplifier: Removes line segments from highly dense paths
      that would not have an impact on their appearance.  Speeds up
      rendering and reduces file sizes.

   ...
 */
```

anntzer replied:
*Sad*, but I can live with that.  Perhaps also mention the issue in a comment?

kbrose replied:
Added a comment. It's possible that there may yet be a simple path forward (pun intended) on this. I'll take a look tomorrow with fresh eyes.

anntzer commented on lib/matplotlib/path.py:
```diff
@@ -206,12 +206,13 @@ def _fast_from_codes_and_verts(cls, verts, codes, internals=None):
-             (self._codes is None or np.all(self._codes <= Path.LINETO)))
+            len(self._vertices) >= 128 and
+            (self._codes is None or np.all(self._codes <= Path.LINETO)) and
+            self._simplify_threshold > 0
```
the test on simplify_threshold should go above (say just after path.simplify), in order to avoid iterating on the codes if possible?  (yes, I realize that's unlikely to be the bottleneck)

also it looks like we may be able to deprecate path.simplify and fold it into path.simplify_threshold being zero or not.

kbrose replied:
Good idea. I think combining the parameters makes a lot of sense. They are definitely redundant in purpose. I'll probably also suggest upping the default from 1/9 to something closer to 1/2. It turns out the simplify_threshold is squared in the simplification code, so 1/9 is actually only reducing paths that are 1/81 pixel away from being not parallel! Not sure what the original logic on that was. Perhaps original author was thinking of it in terms of square pixels difference...?

anntzer replied:
Perhaps that can go into a separate PR (linked to this discussion)?  If you look around mpl/__init__.py and mpl/rcsetup.py you should see how rcparam deprecation is handled.  That'll need a notice in api_changes too (this PR will probably benefit from an entry in the docs whatsnew too).

kbrose replied:
> this PR will probably benefit from an entry in the docs whatsnew too

Yep, already have a draft in `doc/users/whats_new/path_simplification_updates.rst`

kbrose replied:
Realized this morning that it looks like it's squared in simplification code to avoid having to compute sqrt step of norm calculation, so it should be behaving as advertised.

anntzer commented on lib/matplotlib/tests/test_simplification.py:
```diff
@@ -13,11 +15,13 @@
 
-# NOTE: All of these tests assume that path.simplify is set to True
-# (the default)
+# NOTE: All of these tests used to assume that path.simplify is set to True
```
Perhaps just remove the comment?
You may also be able to just use a module-level autouse fixture to set the rcparam (see https://docs.pytest.org/en/latest/fixture.html).

kbrose replied:
thanks for the tip!

kbrose commented:
Things I need help on:

* best way to fix currently failing image comparisons.
* should current style be changed to turn simplification *off* by default to match the current behavior?
    * if not, is it OK to update `path.simplify_threshold` to something larger?
* where/how to document this change? Is the `doc/users/whats_new/path_simplification_updates.rst` enough?

anntzer commented on lib/matplotlib/tests/test_simplification.py:
```diff
@@ -13,11 +15,14 @@
-# NOTE: All of these tests assume that path.simplify is set to True
-# (the default)
+@pytest.fixture
+def set_simplify_and_threshold():
```
Is there a reason *not* to make this autoused? (perhaps, I'm asking)

kbrose replied:
There is exactly one test that turns it off, but it's turned off manually in the test... so no, apart from me not reading pytest docs thoroughly :)

kbrose replied:
Hmm. In fact, using `scope='module'` makes some tests fail. Possibly b/c it's only run once in this module, meaning the test that turns off simplification would effect later tests? Don't know for sure. I'll leave it as is for now.

kbrose replied:
Thanks for all your help so far @anntzer

anntzer commented:
I don't have the time right now to look at the test failures, but will try to come back to it at some point (no guarantees of schedule so if someone else volunteers, even better :-)).
Pinging @tacaswell for default style changes.
doc/users/whats_new/path_simplification_updates.rst is the place to document it.

Overall the PR seems to be heading very much in the right direction, hopefully just needs some polishing.

kbrose commented:
A teaser gif of speedup when manipulating plot with 100000 points:

![simplify](https://user-images.githubusercontent.com/4053170/27319851-1f96f1fc-5559-11e7-87da-a0e79f07aff8.gif)

(I used `byzanz-record` to make gif, don't know why it made Axes background green, looks kinda cool, though...)

kbrose commented:
Anyone know the agg code well? I did some profiling to see if there were any easy wins and it turns out almost all of the time is being spent in `qsort_cells` inside `extern/agg24-svn/include/agg_rasterizer_cells_aa.h`, a quick sort routine...

anntzer commented:
You may have more luck asking on the mailing list(s) (...perhaps).

kbrose commented:
So the most recent batch of commits *should* end up with about 9 image comparison failures. I see two ways to address this:

* Update expected images.
* Add another simplification parameter along the lines of `simplify_antiparallel` that defaults to `False` in `_classic_test.mplstyle`, and controls whether the extra simplification I added is used.

There's obvious reasons to not update expected images, but I also think adding an extra parameter that is really only used during testing is equally bad. Any thoughts?

anntzer commented on src/path_converters.h:
```diff
@@ -595,6 +609,10 @@ class PathSimplifier : protected EmbeddedQueue<9>
             return m_source->vertex(x, y);
         }
 
+        /* NOTE: the below comment is out of date. Lines < 1 pixel in
```
If the comment is out of date, just remove it or update it accordingly.

tacaswell commented:
@mdboom is probably the closest we have to an Agg expert.

I am :+1: on updating those test images.  The changes are very small and the speed is definitely worth it.  I think a better test of this speed up is to use right click with the pan tool and zoom the x-range way in and out.

As a side note, there is a `python tools/test_triage.py` will bring up a little Qt app to make looking at test failures much easier.

tacaswell commented on src/path_converters.h:
```diff
@@ -677,8 +703,9 @@ class PathSimplifier : protected EmbeddedQueue<9>
-                m_lastMax = true;
+                m_dnorm2ForwardMax = m_origdNorm2;
+                m_dnorm2BackwardMax = 0.0;
+                m_lastForwardMax = true;
```
should `m_lastBackwardMax` be reset here as well?

kbrose replied:
I'm pretty sure it can only be false at this point in the code. It wouldn't hurt though, so if you're worried about it I'm OK adding that in.

tacaswell replied:
It is also helpful breadcrumbs for the next person to look at this code.

kbrose replied:
added in latest commit

tacaswell commented on doc/users/whats_new/path_simplification_updates.rst:
```diff
@@ -0,0 +1,8 @@
+``path.simplify_threshold`` RC parameters, as long as other conditions
+such as not using dashed lines are met. Simplification will only happen
+when drawing the line segment portion of a path, if you are also drawing
+markers you should consider using the ``markevery`` option to ``plot``.
```
Can you add a remark here about reasonable values and why you would want to use this?

As near as I can tell this feature is un-documented except for a comment in the change log.

anntzer replied:
There was some discussion as to folding path.simplify and path.simplify_threshold into a single rcparam (zero vs nonzero) that got lost in the thread, too.  @kbrose Do you plan to do it?  It can be a separate PR of course.

kbrose replied:
I'm adding some more text for the whats new entry now. Do you think it's worth adding more (permanent) documentation somewhere?

kbrose replied:
Yeah I think it would be cleaner as a separate PR (I'm willing to do it, but I'll be focusing on this and other performance optimizations first). @anntzer @tacaswell the idea was that `path.simplify` could be removed because it is redundant with `path.simplify_threshold == 0`.

anntzer replied:
No hurry.
Perhaps start a "performance" section in http://matplotlib.org/faq/index.html or http://matplotlib.org/faq/usage_faq.html?

kbrose replied:
I liked that idea, but usage faq has been removed from master? Looks like it happened in eaaa2b4acff07bc6cd3fcf015f6b09f375b2dff9. Guess I'll put somewhere else

tacaswell replied:
It probably got moved to 'tutorials/01_introductory/usage.py' but github is being too lazy about the diffs.

tacaswell commented:
I think I understand this and am :+1: over all.

Might be worth resetting a bit more of the internal state around line 708. If `m_dnorm2BackwardMax == 0` then the values of the back point don't matter, but probably best to reset them anyway.

tacaswell commented:
```python

In [25]: p = mpath.Path(np.array([[0, 1, 3], [0, 0, 0]]).T)

In [26]: p.cleaned(simplify=True)
Out[26]: 
Path(array([[ 0.,  0.],
       [ 3.,  0.],
       [ 3.,  0.],
       [ 0.,  0.]]), array([1, 2, 2, 0], dtype=uint8))
```

This seems to pick up an extra degenerate point at the end, but that is also the behavior of the current code so probably ok to leave that for now.

A recent-ish github feature is ability for maintainers to push commits to PR barnches.  Do you mind if I push some more tests to your branch or would you prefer a PR from me into your branch?

tacaswell commented:
There seems to be a bug if all of the points lie on the vector:

```python
In [60]: p = mpath.Path(np.array([[0, 1, 3, -1, 1], [0, 0, 0, 0, 0]]).T)

In [61]: p.cleaned(simplify=True)
Out[61]: 
Path(array([[ 0.,  0.],
       [ 3.,  0.],
       [ 1.,  0.],
       [ 0.,  0.]]), array([1, 2, 2, 0], dtype=uint8))

```

kbrose commented:
@tacaswell please feel free to push tests directly. I'll look into that behavior.

tacaswell commented:
The tests I just pushed fail in the all-on-the-line case I mentioned above.

kbrose commented:
@tacaswell that's a good catch, and I think I know why it happens. Digging in there also seems to show how those degenerate points are being added: https://github.com/kbrose/matplotlib/blob/updated-downsampling/src/path_converters.h#L794

I'm not sure why that's being done. I'll leave it in for now unless I can figure out why it was added in first place.

kbrose commented:
Most recent commit will end up with some more image failures. Most of them looked pretty small, except the clipping testing. I want to do more testing to be sure, but I think I fixed a long-standing bug that might actually mean the new version is the "correct" version:

```diff
diff --git a/src/path_converters.h b/src/path_converters.h
...
@@ -702,9 +703,12 @@ class PathSimplifier : protected EmbeddedQueue<9>
                 m_dnorm2ForwardMax = m_origdNorm2;
                 m_dnorm2BackwardMax = 0.0;
                 m_lastForwardMax = true;
+                m_lastBackwardMax = false;
 
-                m_nextX = m_lastWrittenX = m_lastx = *x;
-                m_nextY = m_lastWrittenY = m_lasty = *y;
+                m_lastWrittenX = m_lastx;
+                m_lastWrittenY = m_lasty;
+                m_nextX = m_lastx = *x;
+                m_nextY = m_lasty = *y;
                 continue;
             }
 ...
```
The `lastWritten` variables were being set to the *tip* of the new vector in this portion of the code, but everywhere else they were being set to the *base* of the vector. I assume the base is the correct decision. That position is also helped by the fact that this change makes the tests from tacaswell pass (after I removed some the case of `pi` which *should* fail as written, also explained in comment added in commit).

tacaswell commented on lib/matplotlib/tests/test_simplification.py:
```diff
@@ -138,13 +138,18 @@ def _get_simplified(x,y):
+# using min/max will get the expected results out of order:
+# min/max for simplification code depends on original vector,
+# and if angle is outside above range then simplification
+# min/max will be opposite from actual min/max.
```
:+1:

tacaswell commented:
It looks like the svg tests will also need to be updated, do you have `inkscape` installed on your system?

For the images can you squash all of the re-rendered images into one commit (so we don't have images in the history that never match master)?

tacaswell commented:
:+1: I think this is getting very close!

kbrose commented:
> do you have `inkscape` installed

Nope, I'll work on getting that so I can run them.

I'll squash the image commits. Does it matter the ordering? E.g. should I put the squashed commit at the beginning of my changes, the end, doesn't matter?

kbrose commented:
@kbrose TODO: update names of `m_lastWritten*` to something more descriptive, like `m_currentVectorOrigin*` or some such.

anntzer commented:
It's better if you can make all intermediate commits pass the tests (whatever they are at the time of that commit), to help later bisection efforts, but other than that it doesn't really matter.
But we're not very strict on that regard.

tacaswell commented:
It might be best to squash this down to one or two commits all together.  The c++ code is confusing enough without partially working versions in the history.  Can you either just squash it all to few editorialized commits, a single commit, ore we can just use the squash-merge button.

tacaswell commented:
:+1:  I'm pretty excited about this going in!

On a bit more consideration, I think this should be compressed to two commits, 1 for the code changes and one from the image updates.  @kbrose I am happy to do this for you if you want (and force-push to your branch or open a new PR).

Looking at the svg diffs is interesting.  Most of the change is from a formatting change we did a while ago to round the coordinates and strip trailing 0s, but you can see a bunch of lines where are point in a line was eliminated.

kbrose commented:
I can do that rebasing, that's no problem. I'll do another pass through the code to see if there's more commenting to do, and also write something in the docs so it's not just in release notes. I'm away from my computer this weekend, so that'll probably happen Monday.

tacaswell commented:
no problem, have a good weekend!

kbrose commented:
Going to do the rebasing now -- I've put a backup of the branch as it currently exists at https://github.com/kbrose/matplotlib/tree/downsample-backup.

kbrose commented:
@tacaswell I'm pretty happy with how things look. Let me know if there's anything else I can do.

kbrose commented:
For a reproducible test of speed up, I ran 

```
python -m timeit -c "import matplotlib as mpl; mpl.rcParams['path.simplify_threshold'] = 1.0; import matplotlib.pyplot as plt; import numpy as np; plt.plot(np.random.rand(100000)); plt.savefig('test.png')"
```
on the master branch and this branch. Master branch had
```
10 loops, best of 3: 10.6 sec per loop
```
this branch had
```
10 loops, best of 3: 2.22 sec per loop
```
So about 1/5th the rendering time.

kbrose commented:
Previous comment was bad measure of performance since it was diluted by overhead from import statements. I re-ran rendering correctly using `setup` with `timeit` and got more detailed comparisons.

![figure_1](https://user-images.githubusercontent.com/4053170/28152668-4cc42148-6767-11e7-9118-ae911b8e7126.png)

Approximate timing code:
```python
import matplotlib.pyplot as plt
import matplotlib as mpl
import numpy as np
import timeit

mpl.rcParams['path.simplify_threshold'] = 1.0

ts = []
ns = list(range(100000, 1000000, 100000))
for n in ns:
    x = np.random.random(n)
    ts.append(timeit.timeit('plt.plot(x); plt.savefig("test.png")',
                            number=1,
                            globals=globals()))

print(ns)
print(ts)
```

kbrose commented:
@tacaswell Anything more I can do here to get this PR to a place where everyone is comfortable with it?

tacaswell commented:
@kbrose No, at this point you are waiting on us being slow.  If there is no more feedback in a day or two I will merge.

mdboom commented on src/path_converters.h:
```diff
@@ -789,22 +847,47 @@ class PathSimplifier : protected EmbeddedQueue<9>
     inline void _push(double *x, double *y)
     {
-        queue_push(agg::path_cmd_line_to, m_nextX, m_nextY);
+        m_needToPushBack = (m_dnorm2BackwardMax > 0.0);
```
I don't think this needs to be a member variable, as it isn't used from outside this method.

kbrose replied:
Thanks for reviewing. This should be fixed in the new (forcefully rebased) commits.

mdboom commented:
This looks great.  I do worry it might be overoptimizing for one case (a noisy 1-D signal) and *possibly* (though I have no proof) regressing other things.  Can you run your speed comparison on the following?

- A long line with many segments

```
x = np.arange(n)
```

- 2-D noisy signal

```
x = np.random.random(n)
y = np.random.random(n)
```

- A sine wave with many segments

```
x = np.sin(np.arange(n))
```

That would just give me some peace of mind that this isn't inadvertently creating a regression.

kbrose commented:
@mdboom I don't see any significant differences in those test cases. (Sorry, no plots this time, I just compared line by line.)

```
Switched to branch 'master'
master - arange
[100000, 200000, 300000, 400000, 500000, 600000, 700000, 800000, 900000]
[0.1267618790002416, 0.12155359200005478, 0.12315381999997044, 0.13917084200011232, 0.13939967199985404, 0.1533219170000848, 0.16636208199997782, 0.1808252339997125, 0.17434952599978715]
master - 2d random
[10000, 20000, 30000, 40000, 50000, 60000]
[0.8562169729998459, 1.6727452069999345, 2.641577983999923, 3.730134807000013, 4.948208415999943, 6.3215176420003445]
master - sin wave
[100000, 200000, 300000, 400000, 500000, 600000, 700000, 800000, 900000]
[0.1378529929997967, 0.12199795000015001, 0.1291464480000286, 0.1455379730000459, 0.1539971140000489, 0.16803773599986016, 0.18043615000033242, 0.1942135560002498, 0.1985573749998366]

Switched to branch 'updated-downsampling'
updated-downsampling - arange
[100000, 200000, 300000, 400000, 500000, 600000, 700000, 800000, 900000]
[0.12474075899990567, 0.11748132899992925, 0.1221172860000479, 0.13850686100022358, 0.13769903299998987, 0.15148250799984453, 0.16538779599977715, 0.17879632799986211, 0.17407957499972326]
updated-downsampling - 2d random
[10000, 20000, 30000, 40000, 50000, 60000]
[0.8288207969999348, 1.6583962320000865, 2.6092837740002324, 3.6251530230001663, 4.905929156000184, 6.117415098000038]
updated-downsampling - sin wave
[100000, 200000, 300000, 400000, 500000, 600000, 700000, 800000, 900000]
[0.13040111700001944, 0.12126033100003042, 0.1293665680000231, 0.14609414599999582, 0.15306799900008627, 0.1670811919998414, 0.18586396099999547, 0.1956840809998539, 0.19954614300013418]
```

Profiling code:

```bash
# create profile.py
cat > profile.py << EOL
import matplotlib.pyplot as plt
import matplotlib as mpl
import numpy as np
import timeit
import sys

mpl.rcParams['path.simplify_threshold'] = 1.0

ts = []
ns = list(range(int(sys.argv[1]), int(sys.argv[2]), int(sys.argv[3])))
for n in ns:
    ts.append(timeit.timeit(sys.argv[4] + '; plt.savefig("test.png"); plt.close()',
                            number=1,
                            globals=globals()))

print(ns)
print(ts)

EOL

git checkout master
python setup.py build -f > /dev/null 2>&1
pip install . --upgrade > /dev/null 2>&1
echo "master - arange"
python profile.py 100000 1000000 100000 "x = np.arange(n); plt.plot(x)"
echo "master - 2d random"
python profile.py 10000 70000 10000 "x = np.random.random(n); y = np.random.random(n); plt.plot(x, y)"
echo "master - sin wave"
python profile.py 100000 1000000 100000 "x = np.sin(np.linspace(0, 6.25, n)); plt.plot(x)"
echo ""
git checkout updated-downsampling
python setup.py build -f > /dev/null 2>&1
pip install . --upgrade > /dev/null 2>&1
echo "updated-downsampling - arange"
python profile.py 100000 1000000 100000 "x = np.arange(n); plt.plot(x)"
echo "updated-downsampling - 2d random"
python profile.py 10000 70000 10000 "x = np.random.random(n); y = np.random.random(n); plt.plot(x, y)"
echo "updated-downsampling - sin wave"
python profile.py 100000 1000000 100000 "x = np.sin(np.linspace(0, 6.25, n)); plt.plot(x)"
```

mdboom commented:
Cool.  So no significant regressions then.  :+1: from me.

kbrose commented:
After most recent updates, appveyor and circleci fail. I can't tell if this is on my end or not, though. Is there something I need to do (rebase on top of master maybe)?

QuLogic commented:
I'm not understanding some of these test image changes. For example, if I revert `polar_axes.png` to `master` and run this test, they appear almost identical. So instead, I ran it through imagemagick: `compare polar_axes.png polar_axes-expected.png - | display -`:
![diff](https://user-images.githubusercontent.com/302469/28650652-442a950a-724b-11e7-9ab4-a5396e58ca88.png)
which shows (the red highlight) that the majority of the difference is the innermost circle, which is dashed. But if I read the what's new correctly, this code should not affect dashed lines.

For CircleCI, you will need to rebase so that you get the latest config file. I'm not sure what's up with AppVeyor.

kbrose commented:
@QuLogic 

1. Looks like I missed a case where simplification can be applied to dashed lines. In polar plots, a segment inside a dashed line is not guaranteed to be straight, so it could undergo simplification. I'm guessing that's what we're seeing there. It's probably occurring with the inner circle in particular because that contains the most curved line segments inside the dashed line.
2. It looks like appveyor is failing on all recently run CI tests, so I'm guessing this is not just me...
3. I'll rebase on master for the circleCI config.

dstansby commented:
Just restarted appveyor for this PR which should hopefully pass fine.

tacaswell commented:
What is the mechanism for the simplification getting applied to dashed lines in non-linear projections?

kbrose commented:
I assume that each dash is being treated similarly to the standard line and since it's a non-linear projection, a single dash is now represented by > 2 vertices. This is all speculation, though. If anyone has some pointers on where to look to see how dashed lines become curved during radial plotting that would help direct my investigation.

kbrose commented:
@tacaswell It turns out dashed lines are also simplified, and _not_ on a per-dash level. On ***master***, if I run

```python
import matplotlib as mpl
import matplotlib.pyplot as plt
import numpy as np
mpl.rcParams['path.simplify_threshold'] = 1/9
plt.plot(np.random.rand(10000000), '--')
plt.show()
```

it fails with an overflow error, but if I change the simplify threshold to 1.0 then it is able to plot. (Since all the dashes would be exactly straight lines with two vertices, this shows that simplification happens above the per-dash level.)

This poses potential problems: simplifying the anti-parallel case may not produce expected plots when using dashes. On the master branch, the above code (1.0 simplification threshold) produces:

<img width="400" alt="figure_1" src="https://user-images.githubusercontent.com/4053170/28761893-ab6accba-7577-11e7-972c-1f022ed1596e.png">

On this branch, the above code (1.0 simplification threshold) produces :

<img width="400" alt="figure_2" src="https://user-images.githubusercontent.com/4053170/28761896-add1c684-7577-11e7-8c4c-f263a6d3ddf3.png">

But 1.0 simplification threshold is pretty extreme. On this branch, the above code *at 1/9 simplification threshold (default)* produces:

<img width="400" alt="figure_3" src="https://user-images.githubusercontent.com/4053170/28761981-7fe38bbc-7578-11e7-89d8-de629028169a.png">

# Paths forward

* Do nothing (behavior is acceptable)
* Modify more agg code to be able to access the line style and use that to "turn off" anti-parallel simplification
    * The hard part here is updating agg code to be able to get at that. I'm now quite familiar with the simplification parts of the library, but preliminary investigation shows this would involve (at a minimum) touching `src/py_adaptors.h` and `src/_backend_agg.h`, files I am not yet familiar with. Turning off anti-parallel simplification will be a 1 line change after that (change an `else` to an `else if` to make sure `m_dnorm2BackwardMax` always stays 0).
* Give up.

Thoughts?

anntzer commented:
I vote for "behavior is acceptable" with a note in the doc.

kbrose commented:
[Here's](https://gist.github.com/kbrose/36777188944d00ca3fbc6b3cdb0fe4f5) a patch that's like 70% of the way to solution 2. I'm still a lot more uncomfortable with this than the original PR, because this changes several C++ function signatures and adds another attribute to `Path`. It was also complicated by the fact that the linestyle information is not in the `Path` object, but in the `GraphicsContext` (in an attribute which @anntzer just recently deprecated, so I'm playing around with the `get_dashes` function instead), so I'd have to make sure that the path attribute is correctly changed *everywhere* it needs to before being passed to C++ code.

tacaswell commented:
Backing up a bit, why did we think the behavior was to not simplify dashed lines (as it apparently does!)?

It does not make sense to me to enable forward, but not backward, simplification with dashed lines (I assume that the reason only forward was done is that in many case where you are plotting single-valued function with features much bigger than a pixel sampled at densities much smaller than a pixel (ex `sin` with 10k points on a 200 pixel wide axes) the forward direction gets you almost all of the benefit).

I would go with option 1, unless there is a really good reason to skip simplification for dashed lines, in which case 3.

@kbrose Sorry, you have (as is common for non-trivial changes) hit a rabbit hole.  We are hoping there is not a :dragon: at the bottom.

kbrose commented:
TBH I'm not sure where I got that notion in my head. It probably just stuck with me from the PR I opened last year when I was doing simplification in python code (and turned it off for dashed lines). I'm all for going with option 1 (with note in docs) but I actually expected people to be opposed to that option... Shows how much I know :)

anntzer commented:
If you're plotting something so dense that the line keeps overwriting itself, it seems unlikely you'd care about precise dash location (thus my preference for 1).  Note that right now the first line of the whatsnew states exactly the opposite and should be edited.

On an unrelated note, I would also make an additional edit on the (un)suitability of setting the threshold to zero: change
> If you want to make sure your plot reflects your data exactly, then you should set path.simplify to false and/or path.simplify_threshold to 0. Currently, matplotlib defaults to a convervative value of 1/9.

to something like
> If you want to make sure your plot reflects your data exactly, then you should set path.simplify to false and/or path.simplify_threshold to 0. Currently, matplotlib defaults to a convervative value of 1/9 (note that for raster backends, such a value is unlikely to affect the final result in a visible manner).

(Or something else stating that if you're going to rasterize, you already lose a lot of precision anyways.  Compare with https://www.cairographics.org/manual/cairo-cairo-t.html#cairo-set-tolerance.)

tacaswell commented:
option 1 it is, just remove that comment in the whats_new entry and I think we are good to go!

kbrose commented:
@tacaswell Updated documentation and rebased. The actual commit can be found at https://github.com/kbrose/matplotlib/commit/ec755171faa7e9704d3f588b588a804bf1fd6921.

kbrose commented:
Hmm. Docs build failed with unhelpful message

```Warning, treated as error:
The following HTTP Error has occurred: 404
Building HTML failed.
Exited with code 1
```

Building locally seems to give a little more info:

```Exception occurred:
  File "/home/kevin/miniconda3/envs/mpl-dev/lib/python3.5/site-packages/sphinx_gallery/docs_resolv.py", line 52, in _get_data
    with open(url, 'r') as fid:
FileNotFoundError: [Errno 2] No such file or directory: 'https://docs.scipy.org/doc/scipy/reference/searchindex.js'
```

which is strange since I can open the link manually...

QuLogic commented:
Not sure that's the problem link, but please try rebasing against latest `master`.

kbrose commented:
@QuLogic I rebased against master locally without luck. I updated that `_get_data` function to check for http *and* https which appeared to work. I wonder if scipy just switched over to https, and none of the other links that get passed to that function are https so it's not been tested before?

kbrose commented:
Oh, hmm, just realized I updated my install of the `sphinx_gallery` package, not inside matplotlib... That complicates things...

QuLogic commented:
> FileNotFoundError: [Errno 2] No such file or directory: 'https://docs.scipy.org/doc/scipy/reference/searchindex.js'

Oh, this is due to outdated sphinx-gallery; it should work on the latest version.

kbrose commented:
Ok... pip installing up to `0.1.12` fixes that error but adds a new one (building locally, same results on `master` branch):

```Exception occurred:
  File "/home/kevin/miniconda3/envs/mpl-dev/lib/python3.5/site-packages/sphinx_gallery/docs_resolv.py", line 274, in _get_link
    if comb_name in html:
TypeError: a bytes-like object is required, not 'str'
```

Full error:

```
# Sphinx version: 1.6.2
# Python version: 3.5.3 (CPython)
# Docutils version: 0.13.1 release
# Jinja2 version: 2.9.6
...
Traceback (most recent call last):
  File "/home/kevin/miniconda3/envs/mpl-dev/lib/python3.5/site-packages/sphinx/cmdline.py", line 306, in main
    app.build(opts.force_all, filenames)
  File "/home/kevin/miniconda3/envs/mpl-dev/lib/python3.5/site-packages/sphinx/application.py", line 357, in build
    self.emit('build-finished', None)
  File "/home/kevin/miniconda3/envs/mpl-dev/lib/python3.5/site-packages/sphinx/application.py", line 489, in emit
    return self.events.emit(event, self, *args)
  File "/home/kevin/miniconda3/envs/mpl-dev/lib/python3.5/site-packages/sphinx/events.py", line 79, in emit
    results.append(callback(*args))
  File "/home/kevin/miniconda3/envs/mpl-dev/lib/python3.5/site-packages/sphinx_gallery/docs_resolv.py", line 455, in embed_code_links
    _embed_code_links(app, gallery_conf, gallery_dir)
  File "/home/kevin/miniconda3/envs/mpl-dev/lib/python3.5/site-packages/sphinx_gallery/docs_resolv.py", line 388, in _embed_code_links
    full_fname)
  File "/home/kevin/miniconda3/envs/mpl-dev/lib/python3.5/site-packages/sphinx_gallery/docs_resolv.py", line 306, in resolve
    link = self._get_link(cobj)
  File "/home/kevin/miniconda3/envs/mpl-dev/lib/python3.5/site-packages/sphinx_gallery/docs_resolv.py", line 274, in _get_link
    if comb_name in html:
TypeError: a bytes-like object is required, not 'str'```

Are you able to successfully build on master?

QuLogic commented:
You might have some outdated cached data in the doc directory; try deleting all of `doc/build/`, `doc/tutorials`, and `doc/api/_as_gen/`.

kbrose commented:
Looks like rebasing helped. Cleaning docs directory didn't help locally, but I can live with that for now.

QuLogic commented on lib/matplotlib/tests/test_simplification.py:
```diff
@@ -49,33 +51,158 @@ def test_diamond():
+    scale = 5
+    np.random.seed(19680801)
+    # get 15 random offsets
+    # TODO: guarantee offset > 0 results in some offsets < 0
```
This is true at least for the given offsets and current seed:
```python
>>> np.random.seed(19680801)
>>> np.random.rand(15) - 0.5
array([ 0.2003673 ,  0.24275081,  0.20928001,  0.06674552,  0.47778533,
        0.20633485, -0.25208424, -0.34211665,  0.19769852,  0.21995667,
       -0.24225557, -0.15845322,  0.46876117,  0.1945071 , -0.03361674])
```

kbrose replied:
Yep - I put that comment in as a warning to future dabblers

QuLogic commented on tutorials/01_introductory/usage.py:
```diff
@@ -625,3 +625,95 @@ def my_plotter(ax, data1, data2, param_dict):
+#   # Setup, and create the data to plot
+#   y = np.random.rand(100000)
+#   y[50000:] *= 2
+#   y[np.logspace(1,np.log10(50000),400).astype(int)] = -1
```
Please add space after commas.

QuLogic commented on tutorials/01_introductory/usage.py:
```diff
@@ -625,3 +625,95 @@ def my_plotter(ax, data1, data2, param_dict):
+# interactive plotting (with maximal simplification) and another
+# style for publication quality plotting (with minimal
+# simplification) and activate them as necessary. See
+# :ref:`sphx_glr_tutorials_01_introductory_customizing.py` for
```
This link might get changed by #8975, so need to be wary of whichever gets merged first...

kbrose replied:
Note: Current version of that PR would change it to `sphx_glr_tutorials_introductory_customizing.py`

QuLogic commented on tutorials/01_introductory/usage.py:
```diff
@@ -625,3 +625,95 @@ def my_plotter(ax, data1, data2, param_dict):
+#   plt.plot(y)
+#   plt.show()
+#
+#   mpl.rcParams['path.simplify_threshold'] = 1.0
```
This setting is ridiculously fast!

QuLogic commented on tutorials/01_introductory/usage.py:
```diff
@@ -625,3 +625,95 @@ def my_plotter(ax, data1, data2, param_dict):
+#
+# The markevery argument allows for naive subsampling, or an
+# attempt at evenly spaced (along the *x* axis) sampling. See the
+# :ref:`sphx_glr_gallery_pylab_examples_markevery_demo.py`
```
This ref will be outdated by #8983.

kbrose replied:
What's the new ref?

kbrose replied:
`sphx_glr_gallery_lines_bars_and_markers/markevery_demo.py` ?

QuLogic replied:
Yea, that looks right.

QuLogic commented on tutorials/01_introductory/usage.py:
```diff
@@ -625,3 +625,95 @@ def my_plotter(ax, data1, data2, param_dict):
+# controlled by the ``path.simplify`` and
+# ``path.simplify_threshold`` parameters in your
+# ``matplotlibrc`` file (see
+# :ref:`sphx_glr_tutorials_01_introductory_customizing.py` for
```
This link might get changed by #8975, so need to be wary of whichever gets merged first...

kbrose replied:
Note: Current version of that PR would change it to `sphx_glr_tutorials_introductory_customizing.py`

tacaswell commented:
I added a commit to fix the whitespace.  Lets get this in ahead of the two docs PRs if we can.

QuLogic commented:
Too late...

But did you force push something? CircleCI has gone back to its old config for some reason.

kbrose commented:
I guess I'll try rebasing on master again? Also needed up updated docs ref. @tacaswell I'll keep your commit separate for now but if you want that rebased in let me know.

kbrose commented:
Or should I hold off on rebasing since @tacaswell has approved?

QuLogic commented:
@tacaswell's commit is pretty small; I'm sure he won't mind if you rebase and/or squash it.

tacaswell commented:
Feel fee to squash.

kbrose commented:
I meant more that the changes as they exist were already approved, but rebasing makes it harder to verify nothing's been changed since that point. Regardless, I'll rebase to see if that helps the circleCI build. That will probably necessitate also updating the links @QuLogic pointed out.

tacaswell commented:
:tada: Glad this is in!

Thanks and congratulations @kbrose !

kbrose commented:
Woohoo! Thanks for all the help everyone. Looking forward to snappier EKG plotting!
